runtime/lua/vim/cache.lua

```diff
@@ -0,0 +1,481 @@
+-- interop with the native Neovim cache
+if type(package.loaded['vim.cache']) == 'table' then
+  return package.loaded['vim.cache']
+end
+
+local ffi = require('ffi')
+local uv = vim.loop
+
+local M = {}
+
+---@alias CacheHash {mtime: {sec:number, nsec:number}, size:number}
+---@alias CacheEntry {hash:CacheHash, chunk:string}
+
+---@class CacheFindOpts
+---@field rtp? boolean Search for modname in the runtime path (defaults to `true`)
+---@field patterns? string[] Paterns to use (defaults to `{"/init.lua", ".lua"}`)
+---@field paths? string[] Extra paths to search for modname
+
+M.VERSION = 2
+M.path = vim.fn.stdpath('cache') .. '/luac'
+M.enabled = false
+---@type table<string, {total:number, time:number, [string]:number?}?>
+M.stats = {
+  find = { total = 0, time = 0, not_found = 0 },
+}
+
+---@class ModuleCache
+---@field _rtp string[]
+---@field _rtp_pure string[]
+---@field _rtp_key string
+local Cache = {
+  ---@type table<string, table<string,true>>
+  _indexed = {},
+  ---@type table<string, string[]>
+  _topmods = {},
+  _loadfile = loadfile,
```

Comment:
Why is this here? Can't `Cache.loadfile` be use throughout?

Reply:
Sorry, I see `Cache._loadfile` is just `_G.loadfile`. Can this be pulled out?

Reply:
why? I override _G.loadfile with the Cache.loadfile implementation, but I of course still need the original function to load a file when needed

Reply:
I mean just do something like: `local orig_loadfile = _G.loadfile`. There's no need to place this in a table.

Reply:
there's also no harm in it? The whole Cache module is private anyway

Reply:
Ok, leave it for now then. I can go over it all at a later point.